src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+        node_queue[node_queue_size - (RhsIndex)].m_dvec_square_norm)                            \
+    {                                                                                           \
+        const NodeEntry tmp = node_queue[node_queue_size - (LhsIndex)];                         \
+        node_queue[node_queue_size - (LhsIndex)] = node_queue[node_queue_size - (RhsIndex)];    \
+        node_queue[node_queue_size - (RhsIndex)] = tmp;                                         \
+    }
+
+    const size_t NodeQueueSize = 128;
+    NodeEntry node_queue[NodeQueueSize];
+    size_t node_queue_size = 0;
+
+    // Start at the root.
+    node = nodes;
+
+    FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+    while (node->is_interior())
+    {
+        const size_t split_dim = node->get_split_dim();
+        const ValueType split_abs = node->get_split_abs();
+        const ValueType split_dist = query_point[split_dim] - split_abs;
+
+        // Figure out which child node to follow (the one that contains the query point) and which one to push.
+        // Points on the split plane belong to the right child node.
+        const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+        const NodeType* APPLESEED_RESTRICT left_child_node = nodes + node->get_child_node_index();
+        const NodeType* APPLESEED_RESTRICT follow_node = left_child_node + follow_index;
+        const NodeType* APPLESEED_RESTRICT queue_node = left_child_node + 1 - follow_index;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        // Like in the initial step, we stop as soon as we reached a node with enough points.
+        if (follow_node->get_point_count() < IdealLeafSize)
+            break;
+
+        // Push the node that we don't visit now to the node stack.
+        const ValueType square_split_dist = square(split_dist);
+        if (square_split_dist < query_max_square_distance)
+        {
+            VectorType dvec(0.0);
+            dvec[split_dim] = split_dist;
+
+            // Push the node to the node stack.
+            assert(node_queue_size < NodeQueueSize);
+            node_queue[node_queue_size++] = NodeEntry(queue_node, dvec, square_split_dist);
+
+            // Order the top levels of the node stack.
+            if (node_queue_size >= 4)
+            {
+                ORDER_NODE_ENTRIES(4, 3);
+                ORDER_NODE_ENTRIES(2, 1);
+                ORDER_NODE_ENTRIES(4, 2);
+                ORDER_NODE_ENTRIES(3, 1);
+                ORDER_NODE_ENTRIES(3, 2);
+            }
+        }
+
+        // Continue with the child node.
+        node = follow_node;
+    }
+
+    //
+    // Step 4:
+    //
+    //   Visit the nodes in the node stack, from the closest to the farthest,
+    //   updating the stack with new nodes to visit later. We terminate as soon
+    //   as the closest node is farther than our current maximum search distance.
+    //
+
+    while (node_queue_size-- > 0)
```

Comment:
One can do that, but generally I'd recommend splitting the comparison and the decrementation into two lines. Speedwise it shouldn't make any difference (in contrast to a precrement).

Reply:
This is a fairly common C idiom. I used it for compactness, not for performance (as you've noted).

Anyway, for readability's sake, I've split the statement.